db/db_access.c

```diff
@@ -264,6 +269,9 @@ int access_control_check_sql_read(struct BtCursor *pCur, struct sql_thread *thd)
 
     if (pCur->cursor_class == CURSORCLASS_TEMPTABLE)
         return 0;
```

Comment:
This change is good and addresses the issue. However, I think we should also first check for `clnt->authgen != gbl_bpfunc_auth_gen` and invoke `check_sql_access()` on mismatch and then fallback doing the other checks only if `check_sql_access()` succeeds. `check_sql_access()` would take care of setting `clnt->authgen`.

Additionally, an authgen mismatch also implies that we should clear `pCur->permissions` so that it triggers a fresh read/write auth check.

Reply:
So as in check if `clnt->authgen != gbl_bpfunc_auth_gen` at the top of the functions, and if they are not equal, then clear the existing permissions in `pCur->permissions`? If we do this, then we would be able to reauthenticate for the current table, and set `clnt->authgen = gbl_bpfunc_auth_gen` at this point. However, I do not think we would be able to clear the permissions for the rest of the tables since those are in separate `pCur`s. If we didn't set `clnt->authgen = gbl_bpfunc_auth_gen` at this point then we would be able to clear permissions for all of the tables in a query, but would go through the whole procedure for every cursor move again.

Reply:
I don't understand why authgen isn't sufficient and a new flag must be introduced. Seems like you can also bump gbl_bpfunc_auth_gen for revoke/grant, and check per-clnt authgen against gbl_bpfunc_auth_gen for any mismatch?

Reply:
The authgen's are equal in the beginning of a query at this point of the code so then the authentication will always pass if we just check authgens. 

Also if you switch users, change your password, and then run another query, the authgens will be equal again once this part of the code is reached, even though gbl_bpfunc_authgen would be incremented after changing password (since a different part of the code will then set `clnt->authgen = gbl_bpfunc_authgen`), and we run into the same issue as above.

Also I haven't tested this case but I was also thinking what if we run a query that involves two tables, one which we don't have access to. Assume that the authgens are not equal at the start, so we run through `check_sql_access()`. It authenticates for the first table, and then we set `clnt->authgen = gbl_bpfunc_authgen` in the same function. But then we run `check_sql_access()` for the second table, the one we should not have access to. Then I believe we should have `clnt->authgen == gbl_bpfunc_authgen` and will authenticate access when it should fail.

Reply:
And if you have read access to a table but not write access, then select from this table, we will authenticate and set `clnt->authgen = gbl_bpfunc_authgen`. And when we try to insert into this table, we will still have `clnt->authgen == gbl_bpfunc_authgen`

Reply:
One option we could do if we still want to check for permissions in the middle of a query is create an authgen variable within `pCur` and compare that against `gbl_bpfunc_authgen`/update it as needed, but I would still need the permissions variable I also created since I have to check whether read/write access has been granted due to the case above. And would have to increment `gbl_bpfunc_auth_gen` on revoke/grant. Not sure if this is something I should pursue.

Reply:
I think it's reasonable not to recheck permissions in the middle of a query.  If they change, the new permissions will apply to the next query.

Reply:
> I think it's reasonable not to recheck permissions in the middle of a query. If they change, the new permissions will apply to the next query.

I just wanted to keep the old behavior (unsure if that was deliberate design decision). But, I am ok with the idea of not rechecking in the middle of the query execution; checking at every cursor operation seemed like an overkill in first place.

Reply:
> However, I do not think we would be able to clear the permissions for the rest of the tables since those are in separate `pCur`s

Right. However, per last comments, I think the current changes in the PR should be sufficient.

Reply:
> bump gbl_bpfunc_auth_gen for revoke/grant

https://github.com/bloomberg/comdb2/pull/3060